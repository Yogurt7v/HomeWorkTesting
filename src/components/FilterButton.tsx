interface FilterButtonProps {
    buttonName: string;
    onClick: () => void;
}
export function FilterButton({ onClick, buttonName }: FilterButtonProps) {
    return <button className="button button-filter" onClick={onClick}>{buttonName}</button>;
}