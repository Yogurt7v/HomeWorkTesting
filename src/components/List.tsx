import { Item } from "./Item";

type Props = {
  items: Task[];
  onDelete: (id: Task["id"]) => void;
  onToggle: (id: Task["id"]) => void;
};

export const List = ({ items, onDelete, onToggle }: Props) => (

  <ul className="task-list tasks">
    {items.length > 10 && <p>Слишком много задач. Показаны только 10 первых</p>}
    {items.slice(0, 10).map((item) => (
      <Item
        {...item}
        key={item.id}
        onDelete={onDelete}
        onToggle={onToggle} />
    ))}
  </ul>
);
