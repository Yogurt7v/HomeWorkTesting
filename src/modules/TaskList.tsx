import { useDispatch, useSelector } from "react-redux";
import { Empty } from "src/components/Empty";
import { List } from "src/components/List";
import { deleteTask, tasksSelector, toggleTask, filterTusks, unfilteredTusks } from "src/store/taskSlice";
import { FilterButton } from "src/components/FilterButton";
import { useRef, useState } from "react";

export const TaskList = () => {
  const items = useSelector(tasksSelector);
  const dispatch = useDispatch();
  const toogle = useRef(false);
  const [filtered, setFiltered] = useState("Только выполненные");

  const handleDelete = (id: Task["id"]) => {
    dispatch(deleteTask(id));
  };

  const handleToggle = (id: Task["id"]) => {
    dispatch(toggleTask(id));
  };

  const handleFilter = () => {
    if (!toogle.current) {
      dispatch(filterTusks());
      setFiltered("Вернуть все");
      toogle.current = true;
    } else {
      dispatch(unfilteredTusks());
      setFiltered("Только выполненные");
      toogle.current = false;
    }
  }

  return items.length > 0 ? (
    <>
      <FilterButton onClick={handleFilter} buttonName={filtered} />
      <List items={items} onDelete={handleDelete} onToggle={handleToggle} />
    </>

  ) : (
    <>
      <Empty />
    </>
  );
};
