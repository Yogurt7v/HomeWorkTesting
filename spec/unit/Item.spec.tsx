import { render, screen, prettyDOM } from '@testing-library/react';
import { Input } from 'src/components/Input';
import { NewTaskBar } from 'src/modules/NewTaskBar';
import uE from '@testing-library/user-event'
import { JestStoreProvider } from '../utils/JestStoreProvider'
import { List } from 'src/components/List';

const userEvent = uE.setup({ advanceTimers: jest.advanceTimersByTime })


describe('Элемент списка задач', () => {
    it('название не должно быть больше 32 символов', () => {
        render(<Input value='Пример заголовка, содержащего больше 32 символов'
            onChange={() => { }}
            disabled={false}
            disabledMessage=''
        />)
        const labeltext = screen.getByTestId('input-hint-text')
        expect(labeltext.textContent).toBe('Длина заголовка не должна превышать 32 символа')
    });
    it('название не должно быть пустым', async () => {
        render(<NewTaskBar />, {
            wrapper: JestStoreProvider
        })
        const input = screen.getByRole('textbox')
        const addButton = screen.getByRole('button')
        await userEvent.click(input)

        expect(input).toHaveValue('')
        expect(addButton).toBeDisabled()
    });
    it('нельзя удалять невыполненные задачи', () => {
        const fn = jest.fn()
        const mockItems: Task[] = [{
            id: "1",
            header: 'Первая задача',
            done: false,
        },
        {
            id: "2",
            header: 'Вторая задача',
            done: true,
        }
        ]

        render(<List items={mockItems} onDelete={fn} onToggle={() => undefined} />)

        const deleteButton1 = screen.getAllByRole('button')[0]
        const deleteButton2 = screen.getAllByRole('button')[1]

        expect(deleteButton1).toBeDisabled()
        expect(deleteButton2).not.toBeDisabled()

    });
});