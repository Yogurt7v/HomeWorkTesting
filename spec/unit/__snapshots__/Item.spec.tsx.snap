// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Элемент списка задач проверка снэпшота 1`] = `
<DocumentFragment>
  <ul
    class="task-list tasks"
  />
</DocumentFragment>
`;
